src/Strategy/Node/YarnV1/YarnLock.hs

```diff
@@ -0,0 +1,199 @@
+        else dep
+
+    promoteDev = promote EnvDevelopment allDevs
+    promoteProd = promote EnvProduction allProds
+
+mangleParseErr :: FilePath -> YL.LockfileError -> ReadFSErr
+mangleParseErr path = FileParseError path . YL.prettyLockfileError
+
+newtype YarnV1Label
+  = NodeEnvironment DepEnvironment
+  deriving (Eq, Ord, Show)
+
+data YarnV1Package = YarnV1Package
+  { fullPackageName :: Text
+  , packageVersion :: Text
+  }
+  deriving (Eq, Ord, Show)
+
+buildGraph ::
+  forall m sig.
+  ( Has Diagnostics sig m
+  , Has Logger sig m
+  ) =>
+  YL.Lockfile ->
+  FlatDeps ->
+  m (Graphing Dependency)
+buildGraph lockfile flatdeps = withLabeling toDependency $
+  for_ (map firstKey $ MKM.toList lockfile) $ \(key, pkg) -> do
+    let parent :: YarnV1Package
+        parent = pairToPackage key pkg
+
+        keyAsNodePackage :: NodePackage
+        keyAsNodePackage = toNodePackage key
+
+        childrenSpecs :: [YL.PackageKey]
+        childrenSpecs = YL.dependencies pkg
+
+    -- Fetch dependencies and their resolved versions from the lockfile
+    -- Logs a debug message if a map lookup error occurs.
+    children <- catMaybes <$> traverse (resolveVersion lockfile) childrenSpecs
+
+    -- Insert all deps as deep to prevent missing isolated deps.
+    deep parent
+    -- Add edges from current parent
+    traverse_ (edge parent) children
+    -- Mark as direct if present in any relevant package.json direct list
+    if keyAsNodePackage `Set.member` directDeps flatdeps
+      then do
+        direct parent
+        label parent $ NodeEnvironment EnvProduction
+      else pure ()
+    -- Mark as dev if present in any relevant package.json dev list
+    if keyAsNodePackage `Set.member` devDeps flatdeps
+      then do
+        -- We do directly depend on it, even if it's dev.
+        direct parent
+        label parent $ NodeEnvironment EnvDevelopment
+      else pure ()
+
+toDependency :: YarnV1Package -> Set YarnV1Label -> Dependency
+toDependency YarnV1Package{..} = foldr applyLabel start
+  where
+    applyLabel :: YarnV1Label -> Dependency -> Dependency
+    applyLabel (NodeEnvironment env) = insertEnvironment env
+
+    start =
+      Dependency
+        { dependencyType = NodeJSType
+        , dependencyName = fullPackageName
+        , dependencyVersion = Just $ CEq packageVersion
+        , dependencyLocations = []
+        , dependencyEnvironments = []
+        , dependencyTags = mempty
+        }
+
+toNodePackage :: YL.PackageKey -> NodePackage
+toNodePackage key = NodePackage (extractFullName key) $ YL.npmVersionSpec key
+
+resolveVersion :: Has Logger sig m => YL.Lockfile -> YL.PackageKey -> m (Maybe YarnV1Package)
+resolveVersion lockfile key = logMaybePackage key $ pairToPackage key <$> MKM.lookup key lockfile
+
+logMaybePackage :: Has Logger sig m => YL.PackageKey -> Maybe a -> m (Maybe a)
+logMaybePackage key something = do
+  case something of
+    Nothing -> logWarn $ missingResolvedVersionErrorMsg key
```

Comment:
- [ ] Add comment explaining that this shouldn't happen.  However, the types don't guarantee that this is impossible, and due to a third-party bug, we've run into this before.  The warnings should help us identify where this is coming from and fix that issue.